kornia/augmentation/_2d/intensity/random_snow.py

```diff
@@ -0,0 +1,74 @@
+from typing import Any, Dict, Optional, Tuple
+
+import torch
+
+from kornia.augmentation import random_generator as rg
+from kornia.augmentation._2d.intensity.base import IntensityAugmentationBase2D
+from kornia.color import hls_to_rgb, rgb_to_hls
+from kornia.core import Tensor
+from kornia.core.check import KORNIA_CHECK
+
+
+class RandomSnow(IntensityAugmentationBase2D):
+    r"""Generates snow effect on given tensor image or a batch tensor images.
+
+    Args:
+        snow_coefficient: A tuple of floats (lower and upper bound) between 0 and 1 that control
+        the amount of snow to add to the image, the larger value corresponds to the more snow.
+        brightness: A tuple of floats (lower and upper bound) greater than 1 that controls the
+        brightness of the snow.
+        same_on_batch: If True, apply the same transformation to each image in a batch. Default: False.
+        p: Probability of applying the transformation. Default: 0.5.
+        keepdim: Keep the output tensor with the same shape as input. Default: False.
+
+    Shape:
+        - Input: :math:`(C, H, W)` or :math:`(B, C, H, W)`
+        - Output: :math:`(B, C, H, W)`
+
+    Examples:
+        >>> inputs = torch.rand(2, 3, 4, 4)
+        >>> snow = kornia.augmentation.RandomSnow(p=1.0, snow_coefficient=(0.1, 0.6), brightness=(1.0, 5.0))
+        >>> output = snow(inputs)
+        >>> output.shape
+        torch.Size([2, 3, 4, 4])
+    """
+
+    def __init__(
+        self,
+        snow_coefficient: Tuple[float, float] = (0.5, 0.5),
+        brightness: Tuple[float, float] = (2, 2),
+        same_on_batch: bool = False,
+        p: float = 0.5,
+        keepdim: bool = False,
+    ) -> None:
+        super().__init__(p=p, same_on_batch=same_on_batch, keepdim=keepdim)
+        KORNIA_CHECK(all([0 <= el <= 1 for el in snow_coefficient]), "Snow coefficient values must be between 0 and 1.")
+        KORNIA_CHECK(all([1 <= el for el in brightness]), "Brightness values must be greater than 1.")
+
+        self._param_generator = rg.PlainUniformGenerator(
+            (snow_coefficient, "snow_coefficient", 0.5, (0.0, 1.0)), (brightness, "brightness", None, None)
+        )
+
+    def apply_transform(
+        self, input: Tensor, params: Dict[str, Tensor], flags: Dict[str, Any], transform: Optional[Tensor] = None
+    ) -> Tensor:
+        KORNIA_CHECK(input.shape[1] == 3, "Number of color channels should be 3.")
+        KORNIA_CHECK(len(input.shape) in (3, 4), "Wrong input dimension.")
+        if len(input.shape) == 3:
+            input = input[None, :, :, :]
+        input_HLS = rgb_to_hls(input)
+
+        mask = torch.zeros_like(input_HLS)
+        for batch in range(input_HLS.shape[0]):
+            # Retrieve generated parameters
```

Comment:
Why one-by-one? This cannot be done in a batch?

Reply:
Because we need sample different parameters for each distinct image in the batch. If it's possible without iteration, we can do that.

Reply:
I checked the implementation of parameter generator and as far as I understood it does generate tensor with `Size([1])` each time we do `params['parameter'].to(input)`.

Reply:
Okay, I see, it's only if `input` shape is `(2,3,*,*)`.

Reply:
It should generate (B,) size parameters, of which you may convert it to (B,1,1,1) to make the batch operation work.

Reply:
Yes, the generator from the `plain_uniform.py` returns a tensor of parameters of such shape. But, in the `forward` function of `base.py` (`_BasicAugmentationBase`) it invokes this generator in the following way:
```
batch_prob = self.__batch_prob_generator__(batch_shape, self.p, self.p_batch, self.same_on_batch)
to_apply = batch_prob > 0.5
_params = self.generate_parameters(torch.Size((int(to_apply.sum().item()), *batch_shape[1:])))
```
and assigns such value to the `params`. I would understand if it were `batch_shape.shape[0]` (because `batch_shape` is an input shape and first dimension is the number of images in the batch) instead of `(int(to_apply.sum().item()), *batch_shape[1:]))`, which seems randomly generate size of the parameter tensor (I don't fully understand how `self.__batch_prob_generator__` and why is it here).
I also tried printing `params["snow_coefficient"].to(input).shape` locally just by rerunning module with the input of the same shape over and over again and each time it prints different values.

Reply:
I also tried invoking parameter generator directly: `self._param_generator(input.shape[0])` directly in the `apply_transform`, but for some tests `self._param_generator` is not initialized and I don't know why. Moreover, I didn't see you did like that in other augmentations.

Reply:
You do not need to do that. Currently, apply_transform only receive the ones are selected to be transformed. I think if you set p=1, you will get more various sized tensor.

Reply:
So, do I need to do something like `mask[0:snow_coefficient.shape[0], 1, :, :] = torch.where(input_HLS[:, 1, :, :] < snow_coefficient, 1, 0)` in order for broadcasting to work, but transformation will be applied not to the all images in the batch.

Reply:
Okay. With an input of `BCHW`, the `apply_transform` shall obtain a subset of `NCHW` where `N<=B`, along with parameters `snow_coef` and `brightness` that shaped as `N`. Do you mean that you got the wrong shaping here?

If not, you shall directly transform the whole `NCHW` with those paramters. All other parts shall be handled by the framework.

Reply:
I've got it. I just wanted to clarify whether should I sample `NCHW` from `BCHW` randomly, or can I just take `N` first samples from it, or (?)

Reply:
I see

Reply:
Done